app/internal/statistics.py

```diff
@@ -0,0 +1,299 @@
+import datetime
+from typing import Dict, Tuple, Union
+
+from sqlalchemy import and_, or_, func
+from sqlalchemy.orm import Session
+
+from app.database.models import Event, UserEvent
+from app.routers.user import does_user_exist, get_users
+
+
+SUCCESS_STATUS = 0
+ERROR_STATUS = -1
+INVALID_DATE_RANGE = "End date must be later than start date"
+INVALID_USER = "Invalid user id"
+NIN_IN_DAY = 1440
+
+
+def validate_input(db: Session, userid: int, start: datetime.datetime,
+                   end: datetime.datetime)\
+        -> Tuple[bool, str, datetime.datetime, datetime.datetime]:
+    """ 1. input validations:
+            valid userid.
+            end date > start date.
+        2. date range preparation:
+            start: will be at 00:00 on input date (today if None).
+            end: will be at 23:59:59 on input date (today if None).
+    All statistics are based on full days (00:00:00-23:59:59).
+    Args:
+        db: db session.
+        userid: requested user id number for statistics.
+        start: start of date range.
+        end: end of date range.
+    Returns:
+        valid_input: boolean stating if input values are valid.
+        error_text: str (relevant ib case of error).
+        start: start of date range.
+        end: end of date range.
+    """
+    if not does_user_exist(session=db, user_id=userid):
+        return False, INVALID_USER, start, end
+    if start is None:
+        start = datetime.datetime(year=datetime.datetime.now().year,
+                                  month=datetime.datetime.now().month,
+                                  day=datetime.datetime.now().day)
+    else:
+        start = datetime.datetime(year=start.year,
+                                  month=start.month,
+                                  day=start.day)
+    if end is None:
+        end = start + datetime.timedelta(days=1, seconds=-1)
+    else:
+        end = datetime.datetime(year=end.year, month=end.month, day=end.day)\
+              + datetime.timedelta(days=1, seconds=-1)
+    if start >= end:
+        return False, INVALID_DATE_RANGE, start, end
+    return True, "", start, end
+
+
+def get_events_count_stats(db: Session, userid: int, start: datetime.datetime,
+                           end: datetime.datetime)\
+        -> Dict[str, Dict[str, str]]:
+    """ calculate statistics and events relevant for the requested user
+        and the requested date range.
+        all logic is performed in the db.
+    Args:
+        db: db session.
+        userid: requested user id number for statistics.
+        start: start of date range.
+        end: end of date range.
+    Returns: json with
+        events_count_stats: (for requested date range).
+            meetings_for_user: events that the user has.
+            created_by_user: events that the user has created.
+    """
+    return {"events_count_stats":
```

Comment:
Hard to read :(
Also, prefer a shallow structure if possible.

Export the queries to functions.
Also, WDYT about removing the outer "events_count_stats"?

Reply:
I'll definitely export the queries to functions (should have thought about it myself...)
I gave it much thought and still believe that the outer "statistic area name" wrapping each few statistics makes the whole output much more orgenized and clear...
However, if you strongly believe that removing it is better let me know :-)

Reply:
I must say I'm a great fan of "flat is better than nested". From my experience it greatly reduces the amount of bugs and simplifies the use of the system..

Reply:
> I believe splitting it into two functions, or arranging the nesting a bit better, might help read this in just a quick skim-through.
> 
> >

I've rearranged the function - can you please take a look at the new code and advise me if any change is still required? (I'm not sure weather you ment that the "get_events_count_stats" function should call 1 or 2 other functions just for getting the data from the DB)